storage/remote/queue_manager.go

```diff
@@ -302,11 +341,83 @@ func NewQueueManager(reg prometheus.Registerer, metrics *queueManagerMetrics, lo
 	}
 
 	t.watcher = wal.NewWatcher(reg, wal.NewWatcherMetrics(reg), logger, client.Name(), t, walDir)
+	if t.cfg.SendMetadata {
+		t.metadataWatcher = scrape.NewMetadataWatcher(logger, sm, client.Name(), t, t.cfg.MetadataSendInterval)
+	}
 	t.shards = t.newShards()
 
 	return t
 }
 
+// AppendMetadata queues metadata to be sent to the remote storage.
+func (t *QueueManager) AppendMetadata(metadataList []scrape.MetricMetadata) bool {
+	metadataSet := map[scrape.MetricMetadata]struct{}{}
+
+	for _, metric := range metadataList {
+		metadataSet[metric] = struct{}{}
+	}
+
+	metadata := []*prompb.MetricMetadata{}
+	for entry := range metadataSet {
+		metadata = append(metadata, &prompb.MetricMetadata{
+			Metric: entry.Metric,
+			Help:   entry.Help,
+			Type:   metricTypeToMetricTypeProto(entry.Type),
+			Unit:   entry.Unit,
+		})
+	}
+
+	t.sendMetadata(metadata)
+	return true
+}
+
+func (t *QueueManager) sendMetadata(metadata []*prompb.MetricMetadata) {
+	err := t.sendMetadataWithBackoff(metadata)
+
+	if err != nil {
+		level.Error(t.logger).Log("msg", "non-recoverable error while sending metadata", "count", len(metadata), "err", err)
+	}
+}
+
+func (t *QueueManager) sendMetadataWithBackoff(metadata []*prompb.MetricMetadata) error {
```

Comment:
Given how similar this is to `sendWithBackoff`, can we unify them?  Have it accept a `[]bytes` and do the encoding in the caller?

Reply:
That being said its not clear to me we should have the same approach to sending metadata (yet) - this loop retires for ever, and metadata is very much best effort right now.

Reply:
re the unification of `sendWithBackoff`, I'll give it a whirl.

I still think we should back off, the alternative would be to keep retrying with new data using the specified period and that seems inconsistent with the sending of samples. As user, I would expect the retry semantics to be the same.

Reply:
I agree that the retry semantics should remain the same between metadata and samples, especially as these could be in the same message in the future.

Reply:
Done!